state/cluster/badger/mutator.go

```diff
@@ -218,3 +205,74 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	}
 	return nil
 }
+
+// checkDupeTransactionsInUnfinalizedAncestry checks for duplicate transactions in the un-finalized
+// ancestry of the given block, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInUnfinalizedAncestry(block *cluster.Block, includedTransactions map[flow.Identifier]struct{}, finalHeight uint64) ([]flow.Identifier, error) {
+
+	var duplicateTxIDs []flow.Identifier
+	err := fork.TraverseBackward(m.headers, block.Header.ParentID, func(ancestor *flow.Header) error {
+		payload, err := m.payloads.ByBlockID(ancestor.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			txID := tx.ID()
+			_, duplicated := includedTransactions[txID]
+			if duplicated {
+				duplicateTxIDs = append(duplicateTxIDs, txID)
+			}
+		}
+		return nil
+	}, fork.ExcludingHeight(finalHeight))
+
+	return duplicateTxIDs, err
+}
+
+// checkDupeTransactionsInFinalizedAncestry checks for duplicate transactions in the finalized
+// ancestry, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInFinalizedAncestry(includedTransactions map[flow.Identifier]struct{}, minRefHeight, maxRefHeight uint64) ([]flow.Identifier, error) {
+	var duplicatedTxIDs []flow.Identifier
+
+	// Let E be the global transaction expiry constant, measured in blocks
+	//
+	// 1. a cluster block's reference block height is equal to the lowest reference
+	//    block height of all its constituent transactions
+	// 2. a transaction with reference block height T is eligible for inclusion in
+	//    any collection with reference block height C where C<=T<C+E
```

Comment:
⚠️ 
I think the conditions in the code might be potentially inconsistent (?)
* Elsewhere in the code, we include the boundary value https://github.com/onflow/flow-go/blob/c15e68b2ff74b323d21c8836ea1043d15b9603f4/model/flow/constants.go#L21
* while in here, we _exclude_ the boundary value

Is that consistent?

Reply:
The different ranges are talking about inclusion in collections and blocks.

For collections, we set the collection reference height to min(tx.ReferenceBlock.Height) | tx ∈ collection. (The reference heights of transactions in the collection is inclusive at the lower bound.)

For blocks, it's impossible for a block to contain a transaction which references the block it is within, hence why the reference heights of transactions in blocks is exclusive at the lower bound. The number of reference heights which are valid for inclusion in the block is equal to the expiry window (600).

I think it is more that the documentation is confusing than that the code is inconsistent. The "reference height inclusion window" for blocks and collection don't need to be exactly the same (and practically aren't, because of the expiry buffer for collections).